### ISSUE TITLE
[3.x] FreezingException due to ApolloStore

### ISSUE THREAD

**Issue by ychescale9** [apollographql/apollo-android#3357]
**Summary**
There's a `FreezingException` when adding normalized cache. The issue is similar to https://github.com/apollographql/apollo-android/issues/3347.

**Version**
Reproduced on 3.0.0-alpha04, 3.0.0-alpha05.

**Description**
When adding normalized cache (either `withNormalizedCache` or `withStore`), a `FreezingException` is thrown when freezing `ApolloClient`:

```kt
Uncaught Kotlin exception: kotlin.native.concurrent.FreezingException: freezing of com.rickclephas.kmp.nativecoroutines.$nativeSuspend$lambda-2$FUNCTION_REFERENCE$4@29c2788 has failed, first blocker is com.apollographql.apollo3.cache.normalized.internal.DefaultApolloStore@3cad4c8
    at 0   shared                              0x00000001007911f1 kfun:kotlin.Throwable#<init>(kotlin.String?){} + 97 (/opt/buildAgent/work/c5a36d4d82b914cf/kotlin/kotlin-native/runtime/src/main/kotlin/kotlin/Throwable.kt:24:56)
    at 1   shared                              0x000000010078af9d kfun:kotlin.Exception#<init>(kotlin.String?){} + 93 (/opt/buildAgent/work/c5a36d4d82b914cf/kotlin/kotlin-native/runtime/src/main/kotlin/kotlin/Exceptions.kt:23:58)
    at 2   shared                              0x000000010078b1bd kfun:kotlin.RuntimeException#<init>(kotlin.String?){} + 93 (/opt/buildAgent/work/c5a36d4d82b914cf/kotlin/kotlin-native/runtime/src/main/kotlin/kotlin/Exceptions.kt:34:58)
    at 3   shared                              0x00000001007c3bd5 kfun:kotlin.native.concurrent.FreezingException#<init>(kotlin.Any;kotlin.Any){} + 645 (/opt/buildAgent/work/c5a36d4d82b914cf/kotlin/kotlin-native/runtime/src/main/kotlin/kotlin/native/concurrent/Freezing.kt:17:9)
    at 4   shared                              0x00000001007c5119 ThrowFreezingException + 233 (/opt/buildAgent/work/c5a36d4d82b914cf/kotlin/kotlin-native/runtime/src/main/kotlin/kotlin/native/concurrent/Internal.kt:104:15)
    at 5   shared                              0x0000000100979262 FreezeSubgraph + 4882
    at 6   shared                              0x00000001009a5eab Kotlin_Worker_freezeInternal + 27
    at 7   shared                              0x00000001007c3cf0 kfun:kotlin.native.concurrent#freeze__at__0:0(){0§<kotlin.Any?>}0:0 + 64 (/opt/buildAgent/work/c5a36d4d82b914cf/kotlin/kotlin-native/runtime/src/main/kotlin/kotlin/native/concurrent/Freezing.kt:37:2)
    at 8   shared                              0x0000000100bc0a0b kfun:com.rickclephas.kmp.nativecoroutines#nativeSuspend(kotlinx.coroutines.CoroutineScope?;kotlin.coroutines.SuspendFunction0<0:0>){0§<kotlin.Any?>}kotlin.Function2<kotlin.Function2<0:0,kotlin.Unit,kotlin.Unit>,kotlin.Function2<platform.Foundation.NSError,kotlin.Unit,kotlin.Unit>,kotlin.Function0<kotlin.Unit>> + 475 (/Volumes/Projects/Rick Clephas/KMP-NativeCoroutines/kmp-nativecoroutines-core/src/appleMain/kotlin/com/rickclephas/kmp/nativecoroutines/NativeSuspend.kt:43:8)
    at 9   shared                              0x00000001006d4cc7 kfun:dev.johnoreilly.mortycomposekmm.shared.MortyRepository#getCharactersNative(kotlin.Int){}kotlin.Function2<kotlin.Function2<dev.johnoreilly.mortycomposekmm.GetCharactersQuery.Characters,kotlin.Unit,kotlin.Unit>,kotlin.Function2<platform.Foundation.NSError,kotlin.Unit,kotlin.Unit>,kotlin.Function0<kotlin.Unit>> + 327 (/.../MortyComposeKMM/shared/src/commonMain/kotlin/dev/johnoreilly/mortycomposekmm/shared/MortyRepository.kt:<unknown>)
    at 10  shared                              0x000000010070d0fe objc2kotlin.948 + 190 (/<compiler-generated>:1:0)
    at 11  iosApp                              0x00000001000e98c3 $s6iosApp22CharacterListViewModelC15fetchCharactersyyF + 147 (/.../MortyComposeKMM/iosApp/iosApp/Features/Characters/CharacterListViewModel.swift:16:38)
    at 12  iosApp                              0x00000001000df882 $s6iosApp18CharactersListViewV4bodyQrvg7SwiftUI0E0PAEE8onAppear7performQryycSg_tFQOyAgEE15navigationTitleyQrAE18LocalizedStringKeyVFQOyAE0D0Vys5NeverOAE05TupleE0VyAE7ForEachVySaySo21SharedCharacterDetailCGSSAA0cd3RowE0VG_AC08nextPageE033_33644916FC6C7112F8D63ACD51779ECFLLQrvpQOy_Qo_SgtGG_Qo__Qo_yXEfU_yycfU0_ + 146 (/.../MortyComposeKMM/iosApp/iosApp/Features/Characters/CharactersListView.swift:18:22)
```

I updated my [MortyComposeKMM fork](https://github.com/ychescale9/MortyComposeKMM/tree/reproduce-crash-with-NativeCoroutines-integration) to reproduce this.

Looks like the issue is [here](https://github.com/apollographql/apollo-android/blob/b998a34e622a61a02675980b80212d26995d2e3e/apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/internal/ApolloCacheInterceptor.kt#L48).

**Comment by martinbonnin** [comment]
Dang. Is it the same thing where the issue goes away if you remove the `ensureNeverFrozen`? I'll add a test case about this too.

**Comment by ychescale9** [comment]
> Dang. Is it the same thing where the issue goes away if you remove the `ensureNeverFrozen`? I'll add a test case about this too.

I guess so, but I saw this `// The store has a MutableSharedFlow that doesn't like being frozen` so didn't try locally 😄

**Comment by martinbonnin** [comment]
This is from the standard `coroutines` library but maybe `coroutines-mt` handle thing differently. I'll dig into this.

**Comment by ychescale9** [comment]
I want the new memory model :)

**Comment by martinbonnin** [comment]
That sooo much!

**Comment by martinbonnin** [comment]
Fix & tests are there: https://github.com/apollographql/apollo-android/pull/3358

I also relaxed the restriction on `ApolloRequest`. There's just one `ensureNeverFrozen()` left for `BatchingHttpEngine` because it has some mutable state.

**Comment by martinbonnin** [comment]
I was just able to verify this is working with your MortyComposeKMM branch. You'll also need https://github.com/kuuuurt/multiplatform-paging/issues/18 to have paging working on freezed pagers. 🤞 everything is working now... And looking forward to the new memory model 😄

**Comment by ychescale9** [comment]
Thanks! Yeah we only use Apollo at work and generally try to avoid stateful KMP libraries as much as possible for obvious reasons. The world will be a better place once everyone moves to the new mm:)

**Comment by martinbonnin** [comment]
Fixed in `3.0.0-alpha07`
